reference/matrix/fbcsr_kernels.cpp

```diff
@@ -0,0 +1,511 @@
+//    auto a_col_idxs = a->get_const_col_idxs();
+//    auto b_row_ptrs = b->get_const_row_ptrs();
+//    auto b_col_idxs = b->get_const_col_idxs();
+//    for (size_type a_nz = a_row_ptrs[row];
+//         a_nz < size_type(a_row_ptrs[row + 1]); ++a_nz) {
+//        auto a_col = a_col_idxs[a_nz];
+//        auto b_row = a_col;
+//        cols.insert(b_col_idxs + b_row_ptrs[b_row],
+//                    b_col_idxs + b_row_ptrs[b_row + 1]);
+//    }
+//}
+
+
+template <typename ValueType, typename IndexType>
+void spgemm_accumulate_row(map<IndexType, ValueType> &cols,
+                           const matrix::Fbcsr<ValueType, IndexType> *c,
+                           ValueType scale, size_type row) GKO_NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+void spgemm_accumulate_row2(map<IndexType, ValueType> &cols,
+                            const matrix::Fbcsr<ValueType, IndexType> *a,
+                            const matrix::Fbcsr<ValueType, IndexType> *b,
+                            ValueType scale, size_type row) GKO_NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+void spgemm(std::shared_ptr<const ReferenceExecutor> exec,
+            const matrix::Fbcsr<ValueType, IndexType> *a,
+            const matrix::Fbcsr<ValueType, IndexType> *b,
+            matrix::Fbcsr<ValueType, IndexType> *c) GKO_NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+void advanced_spgemm(std::shared_ptr<const ReferenceExecutor> exec,
+                     const matrix::Dense<ValueType> *alpha,
+                     const matrix::Fbcsr<ValueType, IndexType> *a,
+                     const matrix::Fbcsr<ValueType, IndexType> *b,
+                     const matrix::Dense<ValueType> *beta,
+                     const matrix::Fbcsr<ValueType, IndexType> *d,
+                     matrix::Fbcsr<ValueType, IndexType> *c)
+    GKO_NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+void convert_to_dense(const std::shared_ptr<const ReferenceExecutor> exec,
+                      const matrix::Fbcsr<ValueType, IndexType> *const source,
+                      matrix::Dense<ValueType> *const result)
+{
+    const int bs = source->get_block_size();
+    const size_type nbrows = source->get_num_block_rows();
+    const size_type nbcols = source->get_num_block_cols();
+    const IndexType *const row_ptrs = source->get_const_row_ptrs();
+    const IndexType *const col_idxs = source->get_const_col_idxs();
+    const ValueType *const vals = source->get_const_values();
+
+    const gko::blockutils::DenseBlocksView<const ValueType, IndexType> values(
+        vals, bs, bs);
+
+    for (size_type brow = 0; brow < nbrows; ++brow) {
+        for (size_type bcol = 0; bcol < nbcols; ++bcol) {
+            for (int ib = 0; ib < bs; ib++)
+                for (int jb = 0; jb < bs; jb++)
+                    result->at(brow * bs + ib, bcol * bs + jb) =
+                        zero<ValueType>();
```

Comment:
You could use fill_array to initialize with 0, or maybe just loop over the rows and cols of the results once?

Reply:
We visit every block row once, zero out the entire block row of the output, and then write the sparse values to where there are non-zero blocks. I thought that was reasonably efficient given that we will typically not care about high performance for this kernel. I could use `fill_array` I guess, but only if I assume the Dense matrix is stored row-major. Right now this test is independent of that.

Reply:
fill_array might not ideal, since it ignores the stride, and thus might write into the padding. I think actually a few other algorithms make the same error (reference COO SpMV might be one). Dense is always stored row-major, and I doubt that would change any time soon.